openmetrics/src/main/java/io/airlift/openmetrics/MetricsResource.java

```diff
@@ -0,0 +1,340 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.airlift.openmetrics;
+
+import com.google.common.base.Joiner;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.collect.ImmutableList;
+import io.airlift.log.Logger;
+import io.airlift.openmetrics.types.Counter;
+import io.airlift.openmetrics.types.Gauge;
+import io.airlift.openmetrics.types.Metric;
+import io.airlift.openmetrics.types.Summary;
+import io.airlift.stats.CounterStat;
+import io.airlift.stats.TimeDistribution;
+import org.weakref.jmx.MBeanExporter;
+import org.weakref.jmx.ManagedClass;
+
+import javax.inject.Inject;
+import javax.management.InstanceNotFoundException;
+import javax.management.IntrospectionException;
+import javax.management.JMException;
+import javax.management.MBeanAttributeInfo;
+import javax.management.MBeanInfo;
+import javax.management.MBeanServer;
+import javax.management.MalformedObjectNameException;
+import javax.management.ObjectName;
+import javax.management.ReflectionException;
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+@Path("/metrics")
+public class MetricsResource
+{
+    private static final Logger log = Logger.get(MetricsResource.class);
+    private static final String OPENMETRICS_CONTENT_TYPE = "application/openmetrics-text; version=1.0.0; charset=utf-8";
+    private static final String ATTRIBUTE_SEPARATOR = "_ATTRIBUTE_";
+    private static final String TYPE_SEPARATOR = "_TYPE_";
+    private static final String NAME_SEPARATOR = "_NAME_";
+    private static final Pattern METRIC_NAME_PATTERN = Pattern.compile("[[a-zA-Z]][\\w_]*");
+    private static final LoadingCache<String, Boolean> VALID_NAMES_CACHE = CacheBuilder.newBuilder().build(new CacheLoader<String, Boolean>()
```

Comment:
I don't see how this could grow to a large size considering it's logically bound by the amount of jmx mbeans. If a user is attempting to pull so many metrics that this grows to a large size it would likely never actually return the metrics list.